### PR TITLE
Allow for text angle/gradient to be retrieved

### LINKED ISSUE #3836: Detect text rotation without running recognition
As noted in the [documentation ](https://tesseract-ocr.github.io/tessdoc/ImproveQuality.html#rotation--deskewing), Tesseract performs poorly when the page is at an angle (not a multiple of 90 degrees).  This limitation is not problematic from an accuracy standpoint, as Tesseract accurately reports the angle of text lines, so my existing pipeline rotates and re-runs recognition on any image where the angle is significant.  However, this is computationally inefficient as there does not appear to be any way to get the page angle without also running recognition (despite estimating page angle/gradient being [one of the first things calculated](https://github.com/tesseract-ocr/tesseract/blob/60fd2b4abaa9c5c5c42d32db57576bc95d28a78a/src/textord/textord.cpp#L224-L231)). 

Therefore, it would be of significant benefit to be able to get the page angle without running the entire recognition process.  I'll work on a build that does this myself--my initial thought is to add a config option that tells Tesseract to report the page angle and quit early (before recognition) if median line angle is above a user-defined threshold, however let me know if others have thoughts on implementation.

### PR DESCRIPTION
**Pull request by Balearica**
Tesseract already calculates the average gradient (angle) of text lines within `Textord::TextordPage` at present.  The gradient of the text is useful information (as Tesseract performs poorly when the gradient is not [almost] zero), however the average gradient only exists within the [`Textord::TextordPage` function](https://github.com/tesseract-ocr/tesseract/blob/c9895dbad439e5ff42f45ef1d03fe5e8572f2e89/src/textord/textord.cpp#LL222C18-L222C18) at present, with no way for users to access it.  Using the API, getting the gradient currently requires running `Recognize` or `AnalyseLayout` and using the results to manually re-calculate the gradient. 

This PR allows for users to directly retrieve the existing average gradient value calculated in `Textord::TextordPage` using a function named `GetGradient`.  This function can be called any time after `FindLines` has been run.  ~~I also made `FindLines` public so it can be run directly without running `Recognize` or `AnalyseLayout` first (running `AnalyseLayout` would result in paragraph recognition being run twice).~~

I've already used this branch to implement an auto-rotate feature in the latest version of [Tesseract.js](https://github.com/naptha/tesseract.js/) that (unlike adding an auto-rotate pre-processing step) does not negatively impact performance for images without problematic rotation.  A basic script using `GetGradient` is below for demonstrative purposes, along with a test image (named `rotate_image.png` in the code).  Resolves #3836. 

```
#include <tesseract/baseapi.h>
#include <leptonica/allheaders.h>

int main()
{
    tesseract::TessBaseAPI *api = new tesseract::TessBaseAPI();
    // Initialize tesseract-ocr with English, without specifying tessdata path
    if (api->Init(NULL, "eng")) {
        fprintf(stderr, "Could not initialize tesseract.\n");
        exit(1);
    }

    // Open input image with leptonica library
    Pix *image = pixRead("rotate_image.png");
    api->SetImage(image);

    // Find lines, get average gradient
    api->AnalyseLayout();
    float gradient = api->GetGradient();

    printf("Average Gradient: %f\n", gradient);

    // Destroy used object and release memory
    api->End();
    delete api;
    pixDestroy(&image);

    return 0;
}

```

```
Average Gradient: 0.085301
```

### CODE REVIEW

**Comment by amitdo** [comment]
>I also made `FindLines` public so it can be run directly without running `Recognize` or `AnalyseLayout` first (running `AnalyseLayout` would result in paragraph recognition being run twice).

So paragraphs detection will run twice if `AnalyseLayout` is followed by `Recognize`. This looks like a bug.

It can be solved by using this condition:

https://github.com/tesseract-ocr/tesseract/blob/424b17f997363670d187f42c43408c472fe55053/src/api/baseapi.cpp#L903-L907

also in `AnalyseLayout`.

Then we won't need to expose `FindLines`. 

What do you think?

**Review by egorpugin — APPROVED** [review]

**Comment by zdenop** [comment]
Why not use leptonica solution?
 E.g.

```c++
#include <leptonica/allheaders.h>

int main() {
  PIX *pix2;
  l_float32 angle, conf;
  Pix *image = pixRead("rotate_image.png");
  pix2 = pixFindSkewAndDeskew(image, 2, &angle, &conf);
  printf("Skew angle: %7.2f degrees; %6.2f conf\n", angle, conf);
  pixWrite("fixed_rotate_image.png", pix2, IFF_PNG);

  pixDestroy(&image);
  pixDestroy(&pix2);
  return 0;
}
```

**Comment by amitdo** [comment]
>Why not use leptonica solution?

Because it is not necessary.  Tesseract does it anyway.

**Comment by zdenop** [comment]
My understanding is that users want to fix rotation before running OCR. 
This feature will require to use SetImage twice (first to get the angle and then for the corrected image). I guess my proposal will be faster and without the need to touch tesseract API ;-)

**Comment by Balearica** [comment]
> > I also made `FindLines` public so it can be run directly without running `Recognize` or `AnalyseLayout` first (running `AnalyseLayout` would result in paragraph recognition being run twice).
> 
> So paragraph detection will run twice if `AnalyseLayout` is followed by `Recognize`. This looks like a bug.
> 
> It can be solved by using this condition:
> 
> https://github.com/tesseract-ocr/tesseract/blob/424b17f997363670d187f42c43408c472fe55053/src/api/baseapi.cpp#L903-L907
> 
> also in `AnalyseLayout`.
> 
> Then we won't need to expose `FindLines`.
> 
> What do you think?

I think this makes sense conceptually, however if I understand correctly, such a change would impact the results returned by the `AnalyseLayout` API function when run using default settings (`paragraph_text_based` is `true` by default).  I'm always hesitant to advocate for any change that could impact existing code.  If there is opposition to making `FindLines` public I think it would be fine to leave `FindLines` protected and use `AnalyseLayout` instead even if paragraph detection runs multiple times.  I timed these functions using sample image I posted above, and `FindLines` took ~75ms to run while `DetectParagraphs` took ~2ms to run, so the overall impact of this inefficiency appears to be fairly minor.

**Comment by amitdo** [comment]
You are right that my suggestion changes the current behavior,  so it's not a good idea.

I still think we should expose `FindLines` just to make this use case work.

My new suggestion is to choose one of these option:
* Take my previous suggestion, but instead of reusing  `paragraph_text_based`", add a new user configurable variable and use it inside `AnalyseLayout`.
* Use  `AnalyseLayout` without changes. As you said, speed wise the impact of the two calls to `DetectParagraphs` is quite small.

Apart from this small issue, I like the new feature.

**Comment by Balearica** [comment]
@amitdo I changed `FindLines` back to a protected function.  I'm fine with running `AnalyseLayout` as written.

**Comment by amitdo** [comment]
@zdenop,

After reading @Balearica's answer to your question, do you object to merging this PR?

@stweil, can we merge it?

File: src/ccmain/tesseractclass.h [hunk 1]
```diff
@@ -1001,6 +1004,7 @@ class TESS_API Tesseract : public Wordrec {
   int scaled_factor_;
   FCOORD deskew_;
   FCOORD reskew_;
+  float gradient_;
```
Comment (stweil) [inline src/ccmain/tesseractclass.h:1007]: The new member variable should be initialized in the constructor, too.

**Review by stweil — CHANGES_REQUESTED** [review]

File: src/ccmain/tesseractclass.h (hunk 1 above)
Comment (Balearica) [reply src/ccmain/tesseractclass.h]: Edited.

**Comment by Balearica** [comment]
> _First of all_: changing/extending C++-API should be reflected in C-API too.
Good point, I edited the C API to reflect this change.

**Comment by amitdo** [comment]
Regarding semver, the current revision of this PR only **adds** one method to the public API, so the next version should be 5.4.0.

**Comment by wvanrensburg** [comment]
Anyone know when this will get merged? Almost a year now

**Review by egorpugin — APPROVED** [review]

**Review by stweil — APPROVED** [review]
Thank you!

**Pull request merged**